opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.cpp

```diff
@@ -118,8 +144,86 @@ initFromState(const EclipseState& eclState, const Schedule&)
             + "Pa) and the reference temperature (" +  std::to_string(T_ref) + "K)."
             + ezrokhi_msg
             );
+}
+#endif
 
+template<class Scalar>
+void BrineCo2Pvt<Scalar>::
+setNumRegions(std::size_t numRegions)
+{
+    brineReferenceDensity_.resize(numRegions);
+    co2ReferenceDensity_.resize(numRegions);
+    salinity_.resize(numRegions);
+}
+
+template<class Scalar>
+void BrineCo2Pvt<Scalar>::
+setReferenceDensities(unsigned regionIdx,
+                      Scalar rhoRefBrine,
+                      Scalar rhoRefCO2,
+                      Scalar /*rhoRefWater*/)
+{
+    brineReferenceDensity_[regionIdx] = rhoRefBrine;
+    co2ReferenceDensity_[regionIdx] = rhoRefCO2;
+}
+
+template<class Scalar>
+void BrineCo2Pvt<Scalar>::
+setActivityModelSalt(int activityModel)
+{
+    switch (activityModel) {
+    case 1:
+    case 2:
+    case 3:  activityModel_ = activityModel; break;
+    default: OPM_THROW(std::runtime_error, "The salt activity model options are 1, 2 or 3");
+    }
+}
+
+template<class Scalar>
+void BrineCo2Pvt<Scalar>::
+setThermalMixingModel(int thermalMixingModelSalt, int thermalMixingModelLiquid)
+{
+    switch (thermalMixingModelSalt) {
+    case 0: saltMixType_ = Co2StoreConfig::SaltMixingType::NONE; break;
+    case 1: saltMixType_ = Co2StoreConfig::SaltMixingType::MICHAELIDES; break;
+    default: OPM_THROW(std::runtime_error, "The thermal mixing model option for salt are 0 or 1");
+    }
+
+    switch (thermalMixingModelLiquid) {
+    case 0: liquidMixType_ = Co2StoreConfig::LiquidMixingType::NONE; break;
+    case 1: liquidMixType_ = Co2StoreConfig::LiquidMixingType::IDEAL; break;
+    case 2: liquidMixType_ = liquidMixType_ = Co2StoreConfig::LiquidMixingType::DUANSUN;; break;
```

Comment:
Sorry for not reading this more carefully during review.  This statement (case `2`) appears to assign `liquidMixType_` twice without an intervening sequence point.  Is that intentional?  If not, we should remove one of the references to `liquidMixType_` and/or replace it with the intended object.  Might also want to remove one of the terminating semicolons on `DUANSUN`.

Reply:
woops. some shoddy c&p. supposed to be only one liquidMixType_ here, https://github.com/OPM/opm-common/pull/4215